tools/OutlineService/OutlineService/OutlineService.cs

```diff
@@ -414,67 +441,91 @@ public partial class OutlineService : ServiceBase {
       }
 
-      var gatewayInterface = interfacesWithIpv4Gateways.First();
+      // TODO: When we find multiple interfaces with IPv4 gateways, guess which one is active by,
+      //       for example, choosing the one with the lowest device metric.
+      return interfacesWithIpv4Gateways.First();
+    }
 
-      // Though it's unclear how this could happen, ensure the interface has just one IPv4 gateway.
-      if (gatewayInterface.GetIPProperties().GatewayAddresses
-          .Select(g => g.Address)
-          .Where(a => a.AddressFamily == AddressFamily.InterNetwork).Count() > 1) {
-        throw new Exception($"interface {gatewayInterface.Name} has multiple IPv4 gateways");
-      }
+    private void SetGatewayProperties(NetworkInterface gateway) {
+      gatewayIp = GetInterfaceGatewayIpv4(gateway);
+      gatewayInterfaceName = gateway.Name;
+    }
 
-      // OK, we can handle this routing table!
-      gatewayIp = gatewayInterface.GetIPProperties().GatewayAddresses
-          .Select(g => g.Address)
-          .Where(a => a.AddressFamily == AddressFamily.InterNetwork)
-          .First();
-      gatewayInterfaceName = gatewayInterface.Name;
+    private IPAddress GetInterfaceGatewayIpv4(NetworkInterface networkInterface) {
+      // Though it's unclear how an interface could have multiple gateway addresses,
+      // pick the first one.
+      return networkInterface.GetIPProperties().GatewayAddresses
+            .Select(g => g.Address)
+            .Where(a => a.AddressFamily == AddressFamily.InterNetwork)
+            .First();
     }
 
     // TODO: Notify the UI of failures here - they could leave the system in a broken state.
     private void NetworkAddressChanged(object sender, EventArgs evt) {
-      if (proxyIp == null || gatewayIp == null || gatewayInterfaceName == null) {
+      if (proxyIp == null) {
         eventLog.WriteEntry("Network change but Outline is not connected, nothing to do.");
         return;
       }
 
-      var previousGatewayIp = gatewayIp;
-      var previousGatewayInterfaceName = gatewayInterfaceName;
-
+      NetworkInterface newGateway = null;
       try {
-        GetSystemIpv4Gateway();
-
-        if (previousGatewayIp.Equals(gatewayIp) &&
-            previousGatewayInterfaceName == gatewayInterfaceName) {
-          eventLog.WriteEntry("No route change required.");
-          return;
-        }
-
-        eventLog.WriteEntry("Network change: " +
-            $"{previousGatewayIp.ToString()} -> {gatewayIp.ToString()}, " +
-            $"{previousGatewayInterfaceName} -> {gatewayInterfaceName}");
+        newGateway = GetSystemIpv4Gateway();
       } catch (Exception e) {
-        eventLog.WriteEntry($"Unsupported routing table after network change: {e.Message}");
+        if (e is NoDefautlGatewayFoundException) {
+          eventLog.WriteEntry("No network connectivity, disabling IPv4 routing.");
+          StopRoutingIpv4();

```

Comment:
What about IPv6?

Reply:
We're only guaranteed not to have IPv4 connectivity at this point. We do not want to delete our IPv6 routes, as there may be active IPv6 interfaces.

Reply:
Ah, yes.